gobblin-core/src/test/java/gobblin/async/AsyncDataDispatcherTest.java

```diff
@@ -0,0 +1,237 @@
+package gobblin.async;
+
+import java.util.Queue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+
+@Test
+public class AsyncDataDispatcherTest {
+  /**
+   * Test successful data dispatch with 2 writers. No exception
+   */
+  public void testSuccessfulDataDispatch() {
+    TestAsyncDataDispatcher dispatcher = new TestAsyncDataDispatcher();
+    // This should work when there is nothing to process
+    dispatcher.waitForABufferEmptyOccurrence();
+
+    ExecutorService service = Executors.newFixedThreadPool(2);
+    Writer writer1 = new Writer(dispatcher);
+    Writer writer2 = new Writer(dispatcher);
+    Future<?> future1 = service.submit(writer1);
+    Future<?> future2 = service.submit(writer2);
+
+    // Do some process
+    try {
+      Thread.sleep(200);
```

Comment:
Try not to use specific sleeps in tests, as they may fail in slower or more contested systems. In general, it is better to await for a condition with a relatively long timeout (couple seconds). You can use for example `ExponentialBackoff.awaitCondition()`